File: frontend/src/types/deployment.ts
export type DeploymentState = 'running' | 'stopped' | 'partial' | 'error' | 'unknown' | 'dry_run';

export interface HealthResponse {
  status: string;
  service: string;
  dry_run: boolean;
}

export interface CommandResult {
  command: string[];
  cwd: string;
  exit_code: number | null;
  stdout: string;
  stderr: string;
  started_at: string;
  finished_at: string;
  duration_ms: number;
  timed_out: boolean;
  dry_run: boolean;
}

export interface ContainerStatus {
  name: string;
  service: string | null;
  status: string;
  running: boolean;
}

export interface DeploymentSummary {
  id: string;
  name: string;
  description: string;
  path: string;
  status: DeploymentState;
  mode: string;
  supported_actions: string[];
  validation_checks: string[];
  rf_capable: boolean;
  components: string[];
}

export interface DeploymentStatus {
  id: string;
  status: DeploymentState;
  containers: ContainerStatus[];
  checked_at: string;
  command: CommandResult;
  output: string;
}

export interface DeploymentActionResponse {
  id: string;
  action: string;
  status: DeploymentState;
  command: CommandResult;
  message: string;
}

export interface RfStartPayload {
  execute: boolean;
  confirmation_phrase: string;
  operator_note: string;
  requested_duration_seconds: number;
  acknowledgements: {
    legal_authorization_valid: boolean;
    isolation_and_attenuation_verified: boolean;
    channel_and_gain_reviewed: boolean;
    emergency_stop_accessible: boolean;
  };
}

export interface LogsResponse {
  id: string;
  container: string | null;
  tail: number;
  command: CommandResult;
}

File: backend/app/models/deployment.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


DeploymentState = Literal["running", "stopped", "partial", "error", "unknown", "dry_run"]


class ErrorDetail(BaseModel):
    code: str
    message: str
    exit_code: int | None = None
    stderr: str | None = None
    active_scenario: str | None = None


class ErrorResponse(BaseModel):
    detail: ErrorDetail


class HealthResponse(BaseModel):
    status: str = "ok"
    service: str = "lain5g-lab-backend"
    dry_run: bool


class CommandResult(BaseModel):
    command: list[str]
    cwd: str
    exit_code: int | None
    stdout: str = ""
    stderr: str = ""
    started_at: datetime
    finished_at: datetime
    duration_ms: int
    timed_out: bool = False
    dry_run: bool = False


class ContainerStatus(BaseModel):
    name: str
    service: str | None = None
    status: str
    running: bool


class DeploymentSummary(BaseModel):
    id: str
    name: str
    description: str = ""
    path: str
    status: DeploymentState
    mode: str = "simulation"
    supported_actions: list[str]
    validation_checks: list[str] = Field(default_factory=list)
    rf_capable: bool = False
    components: list[str] = Field(default_factory=list)


class DeploymentStatus(BaseModel):
    id: str
    status: DeploymentState
    containers: list[ContainerStatus] = Field(default_factory=list)
    checked_at: datetime
    command: CommandResult
    output: str = ""


class DeploymentActionResponse(BaseModel):
    id: str
    action: str
    status: DeploymentState
    command: CommandResult
    message: str


class RfAcknowledgements(BaseModel):
    legal_authorization_valid: bool = False
    isolation_and_attenuation_verified: bool = False
    channel_and_gain_reviewed: bool = False
    emergency_stop_accessible: bool = False


class RfStartRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    execute: bool = False
    confirmation_phrase: str = ""
    operator_note: str = Field(default="", max_length=240)
    requested_duration_seconds: int = Field(default=60, ge=1, le=600)
    acknowledgements: RfAcknowledgements = Field(default_factory=RfAcknowledgements)

    @model_validator(mode="after")
    def require_execution_guards(self) -> "RfStartRequest":
        if self.execute and not all(self.acknowledgements.model_dump().values()):
            raise ValueError("All RF safety acknowledgements are required")
        if self.execute and len(self.operator_note.strip()) < 3:
            raise ValueError("An operator note is required for RF execution")
        return self


class LogsResponse(BaseModel):
    id: str
    container: str | None = None
    tail: int
    command: CommandResult


class RawPayload(BaseModel):
    data: dict[str, Any]
